web-api/src/storage/postgres.rs

```diff
@@ -307,45 +368,40 @@ impl Database {
         tx: &mut Transaction<'_, Postgres>,
         ids: impl IntoIterator<IntoIter = impl ExactSizeIterator<Item = &DocumentId>>,
     ) -> Result<Vec<ExcerptedDocument>, Error> {
+        let ids = ids.into_iter();
+
         let mut builder = QueryBuilder::new(
-            "SELECT document_id, snippet, is_summarized, properties, tags, is_candidate
+            "SELECT document_id, raw_document, preprocessing_step, properties, tags, is_candidate
             FROM document
             WHERE document_id IN ",
         );
-        let mut ids = ids.into_iter().peekable();
         let mut documents = Vec::with_capacity(ids.len());
-        while ids.peek().is_some() {
-            documents.extend(
-                builder
-                    .reset()
-                    .push_tuple(ids.by_ref().take(Self::BIND_LIMIT))
-                    .build()
-                    .try_map(|row| {
-                        let (id, snippet, is_summarized, Json(properties), tags, is_candidate) =
-                            FromRow::from_row(&row)?;
-                        Ok(ExcerptedDocument {
-                            id,
-                            snippet,
-                            is_summarized,
-                            properties,
-                            tags,
-                            is_candidate,
-                        })
-                    })
-                    .fetch_all(&mut *tx)
-                    .await?,
-            );
+        let mut chunks = IterAsTuple::chunks(Self::BIND_LIMIT, ids);
+        while let Some(ids) = chunks.next() {
+            let mut rows = builder.reset().push_tuple(ids).build().fetch(&mut *tx);
+
+            while let Some(row) = rows.next().await.transpose()? {
```

Comment:
❓ could't we just do a `fetch_all()` and then iterate over the resulting vec, or is this more efficient in terms of allocations?

Reply:
it is one allocation less and processes results while they come in, but in this context that is an irrelevant micro optimizations we don't have to do,  it's mainly there because the code went through multiple iterations and at some in-between point it was more readable

I want async for :cry:, and generators :old_man_yelling_at_cloud: